### DISCUSSION TITLE
[SDK] Viem's `sendEip712Transaction` not working with hoisted Local Account

### DISCUSSION THREAD

**Discussion by cesarenaldi** [zkSync-Community-Hub/zksync-developers#807] [SDKs] [answered]
### Team or Project

Lens Network

### Select the SDK

JavaScript

### Environment

Testnet

### Issue Description

Although mentioned in the TS doc, the `sendEip712Transaction` action seems to not work with hoisted Local Account.

It errors with:

```bash
file://[OMITTED]/node_modules/.pnpm/viem@2.21.53_typescript@5.7.2/node_modules/viem/_esm/zksync/actions/sendEip712Transaction.js:54
        throw new AccountNotFoundError({
              ^

AccountNotFoundError: Could not find an Account to execute with this Action.
Please provide an Account with the `account` argument on the Action, or by supplying an `account` to the Client.

Docs: https://viem.sh/docs/actions/wallet/sendTransaction#account
Version: viem@2.21.53
    at Module.sendEip712Transaction (file://[OMITTED]/node_modules/.pnpm/viem@2.21.53_typescript@5.7.2/node_modules/viem/zksync/actions/sendEip712Transaction.ts:98:11)
    at [OMITTED]/index.ts:32:20
    at ViteNodeRunner.runModule (file://[OMITTED]/node_modules/.pnpm/vite-node@1.6.0/node_modules/vite-node/dist/client.mjs:362:5)
    at ViteNodeRunner.directRequest (file://[OMITTED]/node_modules/.pnpm/vite-node@1.6.0/node_modules/vite-node/dist/client.mjs:346:5)
    at ViteNodeRunner.cachedRequest (file://[OMITTED]/node_modules/.pnpm/vite-node@1.6.0/node_modules/vite-node/dist/client.mjs:189:14)
    at ViteNodeRunner.executeFile (file://[OMITTED]/node_modules/.pnpm/vite-node@1.6.0/node_modules/vite-node/dist/client.mjs:161:12)
    at CAC.run (file://[OMITTED]/node_modules/.pnpm/vite-node@1.6.0/node_modules/vite-node/dist/cli.mjs:91:5) {
  details: undefined,
  docsPath: '/docs/actions/wallet/sendTransaction',
  metaMessages: undefined,
  shortMessage: 'Could not find an Account to execute with this Action.\n' +
    'Please provide an Account with the `account` argument on the Action, or by supplying an `account` to the Client.',
  version: '2.21.53'
}
```

### Expected Behavior

It should sign and send the transaction.

### Code Example

```ts
import { chains } from '@lens-network/sdk/viem';
import { parseAbi, Hex, createWalletClient, http, encodeFunctionData } from 'viem';
import { privateKeyToAccount } from 'viem/accounts';
import { eip712WalletActions, getGeneralPaymasterInput, sendEip712Transaction } from 'viem/zksync';

const chain = chains.testnet;

const account = privateKeyToAccount(process.env.PRIVATE_KEY as Hex);

const walletClient = createWalletClient({
  account,
  chain,
  transport: http(),
}).extend(eip712WalletActions())

const abi = parseAbi([
  "function ping()"
]);

const hash = await sendEip712Transaction(walletClient, {
  to: '0x06a71429d153026a0F3bAdf481c216FDfe2d0629',
  data: encodeFunctionData({
    abi: abi,
    functionName: 'ping'
  }),
  paymaster: '0xE6b37F099439ED1Eb38a97b909E9AD0ad74270Af',
  paymasterInput: getGeneralPaymasterInput({
    innerInput: '0x',
  })
})
```

deps:

```
"@lens-network/sdk": "canary",
"viem": "^2.21.53"
```

**Comment by alhenaai** [comment]
It appears that the error you're encountering with the `sendEip712Transaction` action in Viem is due to the SDK not being able to locate an account to execute the transaction. This is indicated by the `AccountNotFoundError` that you've shared.

Here are a few steps to troubleshoot and potentially resolve this issue:

1. **Ensure Account Initialization**: Make sure that the account is properly initialized and passed to the `walletClient`. The error suggests that the account might not be correctly recognized by the client. Double-check that the `privateKeyToAccount` function is correctly converting your private key into an account object and that the account object is correctly structured.

2. **Client Configuration**: Verify that the `walletClient` is correctly configured with the account. It should be explicitly set in the client configuration as shown in your code snippet. If the account is dynamically determined or changed, ensure it is updated in the client configuration.

3. **Documentation and Examples**: Review the [Viem documentation](https://viem.sh/docs/actions/wallet/sendTransaction#account) specifically regarding how accounts should be provided to actions. There might be additional configuration or steps required that are specific to the testnet environment or the version of Viem you are using.

4. **Update SDK and Dependencies**: Ensure that all dependencies are up to date. Sometimes, bugs fixed in newer versions can resolve such issues.

5. **Debugging and Logs**: Add more logging before the transaction call to verify that the account object contains all necessary properties (like the address). This can help identify if the account is being passed to the transaction function correctly.

If after these checks the issue still persists, it might be beneficial to reach out directly to the Viem support or check if there are any open issues in their repository that might be related to this problem.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by bxpana** [comment]
Will check this out and get back to you

**Comment by hoshiyari420** [comment]
@cesarenaldi https://github.com/ondefy/paymaster-mobile-workaround/blob/07ab86fec551dacd3458d55be6b9d41ae2a33cc5/src/app/page.tsx#L66-L85
Hope this helps!

**Reply by cesarenaldi** [reply]
Hey @hoshiyari420, thank you for taking the time but I am not sure how it answers the question about `sendEip712Transaction` action.

**Reply by hoshiyari420** [reply]
Hi @cesarenaldi, apologies for sharing incorrect link. Although, I couldn't find the right link, I remember working on it. 

Adding account parameter as below in `sendEip712Transaction` should work. For more details: https://viem.sh/docs/clients/wallet#3-consume-wallet-actions-1

```
const hash = await sendEip712Transaction(walletClient, {
  account,
  to: '0x06a71429d153026a0F3bAdf481c216FDfe2d0629',
  data: encodeFunctionData({
    abi: abi,
    functionName: 'ping'
  }),
  paymaster: '0xE6b37F099439ED1Eb38a97b909E9AD0ad74270Af',
  paymasterInput: getGeneralPaymasterInput({
    innerInput: '0x',
  })
```

Hope this helps :)

Edit: But yes, it looks like a bug. Ideally should work with both ways.

**Reply by dutterbutter** [reply]
Thanks @hoshiyari420 🙏 

@cesarenaldi please confirm if the above solution works for you. In the meantime, we will investigate and look to fix the issue as its clear the code example you have provided should work as well. 

Thanks for posting 🙏

**Reply by cesarenaldi** [reply]
@dutterbutter I ended up patching our SDK code in a similar way. The only issue is with the users of our network which might not use our SDK, but I see there is a PR already submitted to address this.

**Comment by danijelTxFusion** [comment, marked as answer]
There is an open [PR](https://github.com/wevm/viem/pull/3109) for fixing account hoisting with `sendEip712Transaction` and `signEip712Transaction` actions.
